KWD/GPIO/src/GPIOKeywordDetector.cpp

```diff
@@ -0,0 +1,238 @@
+                        .d("sampleRate", audioFormat.sampleRateHz));
+        return false;
+    }
+    if (GPIO_COMPATIBLE_SAMPLE_SIZE_IN_BITS != audioFormat.sampleSizeInBits) {
+        ACSDK_ERROR(LX("isAudioFormatCompatibleWithGPIOWWFailed")
+                        .d("reason", "incompatibleSampleSizeInBits")
+                        .d("gpiowwSampleSizeInBits", GPIO_COMPATIBLE_SAMPLE_SIZE_IN_BITS)
+                        .d("sampleSizeInBits", audioFormat.sampleSizeInBits));
+        return false;
+    }
+    if (GPIO_COMPATIBLE_NUM_CHANNELS != audioFormat.numChannels) {
+        ACSDK_ERROR(LX("isAudioFormatCompatibleWithGPIOWWFailed")
+                        .d("reason", "incompatibleNumChannels")
+                        .d("gpiowwNumChannels", GPIO_COMPATIBLE_NUM_CHANNELS)
+                        .d("numChannels", audioFormat.numChannels));
+        return false;
+    }
+    return true;
+}
+
+std::unique_ptr<GPIOKeywordDetector> GPIOKeywordDetector::create(
+        std::shared_ptr<AudioInputStream> stream,
+        avsCommon::utils::AudioFormat audioFormat,
+        std::unordered_set<std::shared_ptr<KeyWordObserverInterface>> keyWordObservers,
+        std::unordered_set<std::shared_ptr<KeyWordDetectorStateObserverInterface>> keyWordDetectorStateObservers,
+        std::chrono::milliseconds msToPushPerIteration)  {
+
+    if (!stream) {
+        ACSDK_ERROR(LX("createFailed").d("reason", "nullStream"));
+        return nullptr;
+    }
+
+    // TODO: ACSDK-249 - Investigate cpu usage of converting bytes between endianness and if it's not too much, do it.
+    if (isByteswappingRequired(audioFormat)) {
+        ACSDK_ERROR(LX("createFailed").d("reason", "endianMismatch"));
+        return nullptr;
+    }
+
+    if (!isAudioFormatCompatibleWithGPIOWW(audioFormat)) {
+        return nullptr;
+    }
+
+    std::unique_ptr<GPIOKeywordDetector> detector(new GPIOKeywordDetector(
+        stream, keyWordObservers, keyWordDetectorStateObservers, audioFormat));
+
+    if (!detector->init()) {
+        ACSDK_ERROR(LX("createFailed").d("reason", "initDetectorFailed"));
+        return nullptr;
+    }
+
+    return detector;
+}
+
+GPIOKeywordDetector::GPIOKeywordDetector(
+    std::shared_ptr<AudioInputStream> stream,
+    std::unordered_set<std::shared_ptr<KeyWordObserverInterface>> keyWordObservers,
+    std::unordered_set<std::shared_ptr<KeyWordDetectorStateObserverInterface>> keyWordDetectorStateObservers,
+    avsCommon::utils::AudioFormat audioFormat,
+    std::chrono::milliseconds msToPushPerIteration) :
+        AbstractKeywordDetector(keyWordObservers, keyWordDetectorStateObservers),
+        m_stream{stream},
+        m_maxSamplesPerPush((audioFormat.sampleRateHz / HERTZ_PER_KILOHERTZ) * msToPushPerIteration.count()) {
+}
+
+GPIOKeywordDetector::~GPIOKeywordDetector() {
+    m_isShuttingDown = true;
+    if (m_detectionThread.joinable())
+        m_detectionThread.join();
```

Comment:
What happens if not joinable()?

Reply:
This is a pretty standard operation in the AVS, I don't know the details to answer your question I am afraid. Do you have any suggestion?